libs/compact-map/test/Test/Compact/KeyMap.hs

```diff
@@ -1,19 +1,58 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
 {-# OPTIONS_GHC -Wno-orphans #-}
 
 module Test.Compact.KeyMap where
 
+-- import Data.Proxy
+-- import Test.QuickCheck.Classes.Base
+
+import Cardano.Prelude (HeapWords (..), ST)
+import qualified Data.Compact.HashMap as HM
 import Data.Compact.KeyMap as KeyMap
+import Data.Compact.SmallArray
+  ( MArray,
+    PArray,
+    fromlist,
+    isize,
+    mindex,
+    mwrite,
+    tolist,
+    withMutArray,
+  )
+import Data.Foldable (foldl')
+import Data.List (sortBy)
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import System.Random (RandomGen, genWord64, mkStdGen)
 import Test.QuickCheck
 import Test.Tasty
+import Test.Tasty.HUnit (testCaseInfo)
 import Test.Tasty.QuickCheck
 
+-- ==============================================================
+
+genKey :: RandomGen b => b -> (Key, b)
+genKey g = (Key w0 w1 w2 w3, g4)
+  where
+    (w0, g1) = genWord64 g
+    (w1, g2) = genWord64 g1
+    (w2, g3) = genWord64 g2
+    (w3, g4) = genWord64 g3
+
 instance Arbitrary Key where
   arbitrary =
     oneof
       [ Key <$> arbitrary <*> arbitrary <*> arbitrary <*> arbitrary,
         Key <$> chooseAny <*> chooseAny <*> chooseAny <*> chooseAny
       ]
 
+instance HM.Keyed Key where
```

Comment:
This would have to go in the file KeyMap.hs where the type Key is defined. Best to keep the Arbitrary instances in a test file. Arbitrary instances are almost always orphans.

Reply:
Yep, that's exactly what I meant.